macrosynergy/management/shape_dfs.py

```diff
@@ -37,154 +43,143 @@ def reduce_df(df: pd.DataFrame, xcats: List[str] = None,  cids: List[str] = None
             filt2 = dfx['real_date'] >= pd.to_datetime(value[0])
             filt3 = dfx['real_date'] <= pd.to_datetime(value[1])
             dfx = dfx[~(filt1 & filt2 & filt3)]
-
-    xcats_in_df = dfx['xcat'].unique()
-    if xcats is None:
-        xcats = sorted(xcats_in_df)
-    else:
-        missing = sorted(set(xcats) - set(xcats_in_df))
-        if len(missing) > 0:
-            print(f'Missing cross sections: {missing}')
-            xcats.remove(missing)
-
+    
+    xcats = data_frame(xcats_in_df, xcats, 'xcats')
     dfx = dfx[dfx['xcat'].isin(xcats)]
 
     if intersect:
+        
         df_uns = dfx.groupby('xcat')['cid'].unique()
         cids_in_df = list(df_uns[0])
         for i in range(1, len(df_uns)):
             cids_in_df = [cid for cid in df_uns[i] if cid in cids_in_df]
-    else:
-        cids_in_df = dfx['cid'].unique()
+    
+    cids = data_frame(cids_in_df, cids, 'cids')
+    dfx = dfx[dfx['cid'].isin(cids)]
 
-    if cids is None:
-        cids = sorted(cids_in_df)
+    ## Would there ever be any duplicates ?
```

Comment:
Theoretically no, however, never trust data, and always better for data in the wild to double check (it isn't that computational expensive, relative to the errors in the output you can get).